developer/src/tike/project/Keyman.Developer.UI.Project.xmlLdmlProjectFileUI.pas

```diff
@@ -0,0 +1,285 @@
+    function TestKeyboardState(FCompiledName: string; FSilent: Boolean): Boolean;
+  public
+    function DoAction(action: TProjectFileAction; FSilent: Boolean): Boolean; override;
+    property Debug: Boolean read GetDebug write SetDebug;
+    property ProjectFile: TxmlLdmlProjectFileAction read GetProjectFile;
+  end;
+
+implementation
+
+uses
+  Winapi.Windows,
+  System.Classes,
+  System.StrUtils,
+  System.SysUtils,
+  System.Variants,
+  Vcl.Dialogs,
+  Vcl.Graphics,
+  Vcl.Controls,
+  dmActionsMain,
+
+  Keyman.Developer.UI.UfrmMessageDlgWithSave,
+  UfrmMain,
+  UfrmEditor,
+  UfrmMDIEditor,
+  UKeymanTargets,
+  UmodWebHttpServer,
+  Keyman.Developer.System.ServerAPI,
+  Keyman.Developer.UI.ServerUI,
+  KeyboardFonts,
+  KeymanDeveloperUtils,
+  KeymanDeveloperOptions,
+  utilsystem;
+
+function TxmlLdmlProjectFileUI.DoAction(action: TProjectFileAction; FSilent: Boolean): Boolean;
+begin
+  case action of
+    pfaCompile:   Result := CompileKeyboard(FSilent);
+    pfaTestKeymanWeb: Result := TestKeymanWeb(FSilent);
+    pfaInstall:   Result := InstallKeyboard;
+    pfaUninstall: Result := UninstallKeyboard;
+//  pfaDebug:     Result := DebugKeyboard(FSilent);
+    pfaClean:      Result := ProjectFile.Clean;
+  else
+    Result := False;
+  end;
+end;
+
+function TxmlLdmlProjectFileUI.CompileKeyboard(FSilent: Boolean): Boolean;
+var
+  FSave: Boolean;
+begin
+  Result := False;
+
+  if ProjectFile.Modified then
+  begin
+    if not FSilent then
+    begin
+      if not FKeymanDeveloperOptions.AutoSaveBeforeCompiling then
+      begin
+        if TfrmMessageDlgWithSave.Execute(
+            'The keyboard file has been modified.  You must save before compiling.'+#13#10#13#10+
+            'Save the keyboard and continue?',
+            'Always save automatically before compiling',
+            '', True, FSave) in [mrNo, mrCancel] then
+          Exit(False);
+        if FSave then
+        begin
+          FKeymanDeveloperOptions.AutoSaveBeforeCompiling := True;
+          FKeymanDeveloperOptions.Write;
+        end;
+      end;
+
+      if not modActionsMain.actFileSave.Execute then Exit;
+    end
+    else
+      Exit;
+  end;
+
+  if not FSilent then
+    frmMessages.DoShowForm;
+
+  Result :=
+    ProjectFile.CompileKeyboard;
+
+  if Result and
+      TServerDebugAPI.Running and
+      TServerDebugAPI.IsKeyboardRegistered(ProjectFile.TargetFileName) and
+      (ProjectFile.Targets * KMWKeymanTargets <> []) then
+    TestKeymanWeb(True);
+end;
+
+function TxmlLdmlProjectFileUI.GetDebug: Boolean;
+begin
+  Result := ProjectFile.Debug;
+end;
+
+function TxmlLdmlProjectFileUI.GetProjectFile: TxmlLdmlProjectFileAction;
+begin
+  Result := FOwner as TxmlLdmlProjectFileAction;
+end;
+
+procedure TxmlLdmlProjectFileUI.SetDebug(const Value: Boolean);
+begin
+  ProjectFile.Debug := Value;
+end;
+
+function TxmlLdmlProjectFileUI.TestKeymanWeb(FSilent: Boolean): Boolean;   // I4409
+(*var
+  FCompiledName: string;
```

Comment:
This one of them multi-line comments of dead-code we usually remove?

Reply:
Granted, I see this is the part where we'd test an _LDML_ keyboard in _Web_, which isn't going to be supported for this version.  But... surely just saying "Hey, just CTRL-F this function from this other class / file when the time comes" should work well enough?  Or do we expect significant differences in how it'd work?

Reply:
This is on my list to address now in #9948.

Reply:
(Oh yes, there will be significant differences in the kmn debugger and ldml debugger -- they will be completely divergent implementations, I expect)